File: backend/modules/shared/infrastructure/audit/src/main/java/com/healthdata/audit/service/mpi/MPIAuditService.java
package com.healthdata.audit.service.mpi;

import com.healthdata.audit.dto.mpi.*;
import com.healthdata.audit.entity.DataQualityIssueEntity;
import com.healthdata.audit.entity.MPIMergeEntity;
import com.healthdata.audit.repository.DataQualityIssueRepository;
import com.healthdata.audit.repository.MPIMergeRepository;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDate;
import java.time.LocalDateTime;
import java.time.LocalTime;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

/**
 * Service for MPI audit operations
 */
@Service
@RequiredArgsConstructor
@Slf4j
public class MPIAuditService {
    
    private final MPIMergeRepository mpiMergeRepository;
    private final DataQualityIssueRepository dataQualityIssueRepository;
    
    /**
     * Get paginated merge history with filtering
     */
    public Page<MPIMergeEvent> getMergeHistory(String tenantId, MPIMergeFilter filter, Pageable pageable) {
        LocalDateTime startDate = filter.getStartDate() != null 
            ? filter.getStartDate().atStartOfDay() : null;
        LocalDateTime endDate = filter.getEndDate() != null 
            ? filter.getEndDate().atTime(LocalTime.MAX) : null;
        
        Page<MPIMergeEntity> entities = mpiMergeRepository.findMergeHistory(
            tenantId,
            filter.getMergeStatus(),
            filter.getValidationStatus(),
            startDate,
            endDate,
            filter.getMergeType(),
            filter.getMinConfidenceScore(),
            filter.getMaxConfidenceScore(),
            pageable
        );
        
        return entities.map(this::toMPIMergeEvent);
    }
    
    /**
     * Get detailed merge information
     */
    public MPIMergeDetail getMergeDetail(String mergeId, String tenantId) {
        MPIMergeEntity merge = mpiMergeRepository.findByMergeIdAndTenantId(mergeId, tenantId)
            .orElseThrow(() -> new RuntimeException("Merge not found: " + mergeId));
        
        return toMPIMergeDetail(merge);
    }
    
    /**
     * Validate a merge
     */
    @Transactional
    public MPIValidationResult validateMerge(String mergeId, String tenantId, MPIValidationRequest request, String validatedBy) {
        MPIMergeEntity merge = mpiMergeRepository.findByMergeIdAndTenantId(mergeId, tenantId)
            .orElseThrow(() -> new RuntimeException("Merge not found: " + mergeId));
        
        merge.setValidationStatus("VALIDATED".equals(request.getValidationOutcome()) ? "VALIDATED" : "VALIDATION_FAILED");
        merge.setValidatedBy(validatedBy);
        merge.setValidatedAt(LocalDateTime.now());
        merge.setValidationNotes(request.getValidationNotes());
        merge.setHasMergeErrors(request.getHasMergeErrors());
        merge.setHasDataQualityIssues(request.getHasDataQualityIssues());
        merge.setDataQualityAssessment(request.getDataQualityAssessment());
        
        if ("VALIDATED".equals(request.getValidationOutcome())) {
            merge.setMergeStatus("VALIDATED");
        }
        
        mpiMergeRepository.save(merge);
        
        log.info("Merge {} validated by {} with outcome: {}", mergeId, validatedBy, request.getValidationOutcome());
        
        return MPIValidationResult.builder()
            .mergeId(mergeId)
            .validationOutcome(request.getValidationOutcome())
            .validatedBy(validatedBy)
            .validatedAt(LocalDateTime.now())
            .success(true)
            .message("Merge validation recorded successfully")
            .build();
    }
    
    /**
     * Rollback a merge
     */
    @Transactional
    public MPIRollbackResult rollbackMerge(String mergeId, String tenantId, MPIRollbackRequest request, String rolledBackBy) {
        MPIMergeEntity merge = mpiMergeRepository.findByMergeIdAndTenantId(mergeId, tenantId)
            .orElseThrow(() -> new RuntimeException("Merge not found: " + mergeId));
        
        merge.setMergeStatus("ROLLED_BACK");
        merge.setRollbackReason(request.getRollbackReason());
        merge.setRolledBackBy(rolledBackBy);
        merge.setRolledBackAt(LocalDateTime.now());
        
        mpiMergeRepository.save(merge);
        
        log.info("Merge {} rolled back by {} for reason: {}", mergeId, rolledBackBy, request.getRollbackReason());
        
        // In a real implementation, this would trigger actual patient record restoration
        String restoredSourcePatientId = request.getRecreateSourcePatient() ? merge.getSourcePatientId() : null;
        String updatedTargetPatientId = request.getPreserveTargetPatient() ? merge.getTargetPatientId() : null;
        
        return MPIRollbackResult.builder()
            .mergeId(mergeId)
            .rollbackStatus("ROLLED_BACK")
            .rolledBackBy(rolledBackBy)
            .rolledBackAt(LocalDateTime.now())
            .success(true)
            .message("Merge rolled back successfully")
            .restoredSourcePatientId(restoredSourcePatientId)
            .updatedTargetPatientId(updatedTargetPatientId)
            .build();
    }
    
    /**
     * Get data quality issues
     */
    public Page<DataQualityIssueDTO> getDataQualityIssues(String tenantId, String status, String severity, Pageable pageable) {
        Page<DataQualityIssueEntity> entities = dataQualityIssueRepository.findIssuesByTenant(
            tenantId, status, severity, pageable
        );
        
        return entities.map(this::toDataQualityIssueDTO);
    }
    
    /**
     * Resolve a data quality issue
     */
    @Transactional
    public DataQualityIssueDTO resolveDataQualityIssue(String issueId, String tenantId, DataQualityResolveRequest request, String resolvedBy) {
        DataQualityIssueEntity issue = dataQualityIssueRepository.findByIssueIdAndTenantId(issueId, tenantId)
            .orElseThrow(() -> new RuntimeException("Data quality issue not found: " + issueId));
        
        issue.setStatus("RESOLVED");
        issue.setResolvedBy(resolvedBy);
        issue.setResolvedAt(LocalDateTime.now());
        issue.setResolutionNotes(request.getResolutionNotes());
        
        dataQualityIssueRepository.save(issue);
        
        log.info("Data quality issue {} resolved by {} with action: {}", issueId, resolvedBy, request.getResolutionAction());
        
        return toDataQualityIssueDTO(issue);
    }
    
    /**
     * Get MPI metrics
     */
    public MPIMetrics getMetrics(String tenantId, LocalDate startDate, LocalDate endDate) {
        Long totalMerges = mpiMergeRepository.countByTenantIdAndStatus(tenantId, null);
        Long validatedMerges = mpiMergeRepository.countByTenantIdAndValidationStatus(tenantId, "VALIDATED");
        Long rolledBackMerges = mpiMergeRepository.countByTenantIdAndStatus(tenantId, "ROLLED_BACK");
        Long pendingValidation = mpiMergeRepository.countByTenantIdAndValidationStatus(tenantId, "NOT_VALIDATED");
        
        Double averageConfidence = mpiMergeRepository.getAverageConfidenceScore(tenantId);
        
        Double validationRate = totalMerges > 0 ? (validatedMerges.doubleValue() / totalMerges) : 0.0;
        Double rollbackRate = totalMerges > 0 ? (rolledBackMerges.doubleValue() / totalMerges) : 0.0;
        
        Long totalIssues = dataQualityIssueRepository.countByTenantIdAndStatus(tenantId, null);
        Long criticalIssues = dataQualityIssueRepository.countByTenantIdAndSeverity(tenantId, "CRITICAL");
        Long resolvedIssues = dataQualityIssueRepository.countByTenantIdAndStatus(tenantId, "RESOLVED");
        Double resolutionRate = totalIssues > 0 ? (resolvedIssues.doubleValue() / totalIssues) : 0.0;
        
        return MPIMetrics.builder()
            .totalMerges(totalMerges)
            .validatedMerges(validatedMerges)
            .rolledBackMerges(rolledBackMerges)
            .pendingValidation(pendingValidation)
            .validationRate(validationRate)
            .rollbackRate(rollbackRate)
            .averageConfidenceScore(averageConfidence != null ? averageConfidence : 0.0)
            .averageValidationTimeMinutes(0L)  // Would require time-based calculations
            .mergeTypeDistribution(MPIMetrics.MergeTypeDistribution.builder()
                .automatic(0L)
                .manual(0L)
                .assisted(0L)
                .build())
            .dataQualityMetrics(MPIMetrics.DataQualityMetrics.builder()
                .totalIssues(totalIssues)
                .criticalIssues(criticalIssues)
                .resolvedIssues(resolvedIssues)
                .resolutionRate(resolutionRate)
                .build())
            .build();
    }
    
    /**
     * Get accuracy trends
     */
    public MPITrendData getAccuracyTrends(String tenantId, int days) {
        LocalDate endDate = LocalDate.now();
        LocalDate startDate = endDate.minusDays(days);
        
        List<MPIMergeEntity> merges = mpiMergeRepository.findByTenantIdAndDateRange(
            tenantId,
            startDate.atStartOfDay(),
            endDate.atTime(LocalTime.MAX)
        );
        
        Map<LocalDate, List<MPIMergeEntity>> mergesByDate = merges.stream()
            .collect(Collectors.groupingBy(m -> m.getMergeTimestamp().toLocalDate()));
        
        List<MPITrendData.DailyMergeTrend> dailyTrends = new ArrayList<>();
        LocalDate current = startDate;
        while (!current.isAfter(endDate)) {
            List<MPIMergeEntity> dayMerges = mergesByDate.getOrDefault(current, List.of());
            long totalMerges = dayMerges.size();
            long validated = dayMerges.stream().filter(m -> "VALIDATED".equals(m.getValidationStatus())).count();
            long rolledBack = dayMerges.stream().filter(m -> "ROLLED_BACK".equals(m.getMergeStatus())).count();
            double validationRate = totalMerges > 0 ? (validated / (double) totalMerges) : 0.0;
            double avgConfidence = dayMerges.stream()
                .filter(m -> m.getConfidenceScore() != null)
                .mapToDouble(MPIMergeEntity::getConfidenceScore)
                .average()
                .orElse(0.0);
            
            dailyTrends.add(MPITrendData.DailyMergeTrend.builder()
                .date(current.toString())
                .totalMerges(totalMerges)
                .validatedMerges(validated)
                .rolledBackMerges(rolledBack)
                .validationRate(validationRate)
                .averageConfidence(avgConfidence)
                .build());
            
            current = current.plusDays(1);
        }
        
        double avgValidationRate = dailyTrends.stream()
            .mapToDouble(MPITrendData.DailyMergeTrend::getValidationRate)
            .average()
            .orElse(0.0);
        
        double avgRollbackRate = dailyTrends.stream()
            .filter(t -> t.getTotalMerges() > 0)
            .mapToDouble(t -> t.getRolledBackMerges() / (double) t.getTotalMerges())
            .average()
            .orElse(0.0);
        
        return MPITrendData.builder()
            .dailyTrends(dailyTrends)
            .averageValidationRate(avgValidationRate)
            .averageRollbackRate(avgRollbackRate)
            .build();
    }
    
    // DTO conversion methods
    
    private MPIMergeEvent toMPIMergeEvent(MPIMergeEntity entity) {
        // Count data quality issues for this merge
        Long issueCount = dataQualityIssueRepository.countByTenantIdAndStatus(entity.getTenantId(), "OPEN");
        
        return MPIMergeEvent.builder()
            .mergeId(entity.getId().toString())
            .tenantId(entity.getTenantId())
            .sourcePatientId(entity.getSourcePatientId())
            .targetPatientId(entity.getTargetPatientId())
            .mergeType(entity.getMergeType())
            .confidenceScore(entity.getConfidenceScore())
            .mergeStatus(entity.getMergeStatus())
            .validationStatus(entity.getValidationStatus())
            .mergeTimestamp(entity.getMergeTimestamp())
            .performedBy(entity.getPerformedBy())
            .dataQualityIssueCount(issueCount != null ? issueCount.intValue() : 0)
            .matchedAttributes(extractMatchedAttributes(entity))
            .priority(determinePriority(entity))
            .build();
    }
    
    private MPIMergeDetail toMPIMergeDetail(MPIMergeEntity entity) {
        List<DataQualityIssueEntity> issues = dataQualityIssueRepository
            .findByTenantIdAndPatientIdOrderByDetectedAtDesc(
                entity.getTenantId(),
                entity.getTargetPatientId()
            );
        
        return MPIMergeDetail.builder()
            .mergeId(entity.getId().toString())
            .tenantId(entity.getTenantId())
            .sourcePatient(buildPatientSnapshot(entity.getSourcePatientSnapshot()))
            .targetPatient(buildPatientSnapshot(entity.getTargetPatientSnapshot()))
            .mergedPatient(buildPatientSnapshot(entity.getMergedPatientSnapshot()))
            .mergeType(entity.getMergeType())
            .confidenceScore(entity.getConfidenceScore())
            .mergeStatus(entity.getMergeStatus())
            .validationStatus(entity.getValidationStatus())
            .mergeTimestamp(entity.getMergeTimestamp())
            .performedBy(entity.getPerformedBy())
            .matchedAttributes(buildAttributeMatches(entity.getMatchingDetails()))
            .conflicts(new ArrayList<>())
            .matchingAlgorithmDetails(entity.getMatchingDetails())
            .dataQualityIssues(issues.stream()
                .map(this::toDataQualityIssue)
                .collect(Collectors.toList()))
            .overallDataQualityScore(entity.getDataQualityAssessment())
            .validatedBy(entity.getValidatedBy())
            .validatedAt(entity.getValidatedAt())
            .validationNotes(entity.getValidationNotes())
            .rollbackReason(entity.getRollbackReason())
            .rolledBackAt(entity.getRolledBackAt())
            .rolledBackBy(entity.getRolledBackBy())
            .build();
    }
    
    private DataQualityIssueDTO toDataQualityIssueDTO(DataQualityIssueEntity entity) {
        return DataQualityIssueDTO.builder()
            .issueId(entity.getId().toString())
            .tenantId(entity.getTenantId())
            .patientId(entity.getPatientId())
            .issueType(entity.getIssueType())
            .severity(entity.getSeverity())
            .status(entity.getStatus())
            .description(entity.getDescription())
            .affectedField(entity.getAffectedField())
            .currentValue(entity.getCurrentValue())
            .suggestedValue(entity.getSuggestedValue())
            .recommendation(entity.getRecommendation())
            .detectedAt(entity.getDetectedAt())
            .resolvedAt(entity.getResolvedAt())
            .resolvedBy(entity.getResolvedBy())
            .resolutionNotes(entity.getResolutionNotes())
            .build();
    }
    
    private MPIMergeDetail.DataQualityIssue toDataQualityIssue(DataQualityIssueEntity entity) {
        return MPIMergeDetail.DataQualityIssue.builder()
            .issueId(entity.getId().toString())
            .issueType(entity.getIssueType())
            .severity(entity.getSeverity())
            .description(entity.getDescription())
            .affectedField(entity.getAffectedField())
            .recommendation(entity.getRecommendation())
            .build();
    }
    
    private List<String> extractMatchedAttributes(MPIMergeEntity entity) {
        if (entity.getMatchingDetails() == null) {
            return List.of();
        }
        // Extract matched attribute names from matching details JSON
        return entity.getMatchingDetails().keySet().stream()
            .filter(key -> key.startsWith("match_"))
            .collect(Collectors.toList());
    }
    
    private String determinePriority(MPIMergeEntity entity) {
        if (entity.getConfidenceScore() == null) return "MEDIUM";
        
        double confidence = entity.getConfidenceScore();
        if (confidence < 0.7) return "CRITICAL";
        if (confidence < 0.8) return "HIGH";
        if (confidence < 0.9) return "MEDIUM";
        return "LOW";
    }
    
    private MPIMergeDetail.PatientSnapshot buildPatientSnapshot(Map<String, Object> snapshot) {
        if (snapshot == null) {
            return null;
        }
        
        return MPIMergeDetail.PatientSnapshot.builder()
            .patientId((String) snapshot.get("patientId"))
            .firstName((String) snapshot.get("firstName"))
            .lastName((String) snapshot.get("lastName"))
            .dateOfBirth((String) snapshot.get("dateOfBirth"))
            .gender((String) snapshot.get("gender"))
            .ssn((String) snapshot.get("ssn"))
            .mrn((String) snapshot.get("mrn"))
            .build();
    }
    
    private List<MPIMergeDetail.AttributeMatch> buildAttributeMatches(Map<String, Object> matchingDetails) {
        if (matchingDetails == null) {
            return List.of();
        }
        
        // Build attribute matches from matching details JSON
        return matchingDetails.entrySet().stream()
            .filter(e -> e.getKey().startsWith("match_"))
            .map(e -> MPIMergeDetail.AttributeMatch.builder()
                .attributeName(e.getKey().replace("match_", ""))
                .matchScore(0.0)
                .matchType("EXACT")
                .build())
            .collect(Collectors.toList());
    }
}

File: frontend/src/app/services/audit/mpi-audit.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../../environments/environment';

export interface MPIMergeFilter {
  mergeStatus?: string;
  validationStatus?: string;
  startDate?: string;
  endDate?: string;
  mergeType?: string;
  minConfidenceScore?: number;
  maxConfidenceScore?: number;
}

export interface MPIMergeEvent {
  mergeId: string;
  sourcePatientId: string;
  targetPatientId: string;
  confidenceScore: number;
  mergeStatus: string;
  validationStatus: string;
  mergeTimestamp: string;
  dataQualityIssueCount: number;
  priority: string;
}

export interface MPIMergeDetail {
  mergeId: string;
  mergeStatus: string;
  validationStatus: string;
  confidenceScore: number;
  sourcePatientSnapshot: any;
  targetPatientSnapshot: any;
  mergedPatientSnapshot: any;
  matchingAlgorithmDetails: any;
  attributeMatches: any[];
  attributeConflicts: any[];
  dataQualityIssues: any[];
  validationHistory: any[];
}

export interface MPIValidationRequest {
  validationOutcome: string;
  validationNotes: string;
  hasMergeErrors?: boolean;
  hasDataQualityIssues?: boolean;
  dataQualityAssessment?: string;
}

export interface MPIValidationResult {
  mergeId: string;
  validationOutcome: string;
  validatedBy: string;
  validatedAt: string;
  success: boolean;
  message: string;
}

export interface MPIRollbackRequest {
  rollbackReason: string;
  recreateSourcePatient?: boolean;
  preserveTargetPatient?: boolean;
  rollbackStrategy?: string;
}

export interface MPIRollbackResult {
  mergeId: string;
  rollbackStatus: string;
  rolledBackBy: string;
  rolledBackAt: string;
  success: boolean;
  message: string;
  restoredSourcePatientId?: string;
  updatedTargetPatientId?: string;
}

export interface DataQualityIssueDTO {
  issueId: string;
  patientId: string;
  issueType: string;
  severity: string;
  status: string;
  affectedField: string;
  currentValue: string;
  suggestedValue?: string;
  recommendation: string;
  detectedAt: string;
}

export interface DataQualityResolveRequest {
  resolutionAction: string;
  correctedValue?: string;
  resolutionNotes: string;
}

export interface MPIMetrics {
  totalMerges: number;
  validatedMerges: number;
  rolledBackMerges: number;
  pendingValidation: number;
  validationRate: number;
  rollbackRate: number;
  averageConfidenceScore: number;
  averageValidationTimeMinutes: number;
  mergeTypeDistribution: any;
  dataQualityMetrics: any;
}

export interface MPITrendData {
  dailyTrends: any[];
  averageValidationRate: number;
  averageRollbackRate: number;
}

export interface PageResponse<T> {
  content: T[];
  totalElements: number;
  totalPages: number;
  size: number;
  number: number;
}

/**
 * Service for MPI Audit API operations
 */
@Injectable({
  providedIn: 'root'
})
export class MpiAuditService {
  private readonly baseUrl = `${environment.apiUrl}/api/v1/mpi`;

  constructor(private http: HttpClient) {}

  /**
   * Get MPI merge history with filtering
   */
  getMergeHistory(
    filter: MPIMergeFilter = {},
    page: number = 0,
    size: number = 20,
    sort: string = 'mergeTimestamp,desc'
  ): Observable<PageResponse<MPIMergeEvent>> {
    let params = new HttpParams()
      .set('page', page.toString())
      .set('size', size.toString())
      .set('sort', sort);

    if (filter.mergeStatus) params = params.set('mergeStatus', filter.mergeStatus);
    if (filter.validationStatus) params = params.set('validationStatus', filter.validationStatus);
    if (filter.startDate) params = params.set('startDate', filter.startDate);
    if (filter.endDate) params = params.set('endDate', filter.endDate);
    if (filter.mergeType) params = params.set('mergeType', filter.mergeType);
    if (filter.minConfidenceScore !== undefined) {
      params = params.set('minConfidenceScore', filter.minConfidenceScore.toString());
    }
    if (filter.maxConfidenceScore !== undefined) {
      params = params.set('maxConfidenceScore', filter.maxConfidenceScore.toString());
    }

    return this.http.get<PageResponse<MPIMergeEvent>>(`${this.baseUrl}/merges`, { params });
  }

  /**
   * Get detailed information about a specific merge
   */
  getMergeDetail(mergeId: string): Observable<MPIMergeDetail> {
    return this.http.get<MPIMergeDetail>(`${this.baseUrl}/merges/${mergeId}`);
  }

  /**
   * Validate a merge operation
   */
  validateMerge(mergeId: string, request: MPIValidationRequest): Observable<MPIValidationResult> {
    return this.http.post<MPIValidationResult>(`${this.baseUrl}/merges/${mergeId}/validate`, request);
  }

  /**
   * Rollback a merge operation
   */
  rollbackMerge(mergeId: string, request: MPIRollbackRequest): Observable<MPIRollbackResult> {
    return this.http.post<MPIRollbackResult>(`${this.baseUrl}/merges/${mergeId}/rollback`, request);
  }

  /**
   * Get data quality issues with filtering
   */
  getDataQualityIssues(
    status?: string,
    severity?: string,
    page: number = 0,
    size: number = 20
  ): Observable<PageResponse<DataQualityIssueDTO>> {
    let params = new HttpParams()
      .set('page', page.toString())
      .set('size', size.toString());

    if (status) params = params.set('status', status);
    if (severity) params = params.set('severity', severity);

    return this.http.get<PageResponse<DataQualityIssueDTO>>(`${this.baseUrl}/data-quality/issues`, { params });
  }

  /**
   * Resolve a data quality issue
   */
  resolveDataQualityIssue(issueId: string, request: DataQualityResolveRequest): Observable<DataQualityIssueDTO> {
    return this.http.post<DataQualityIssueDTO>(`${this.baseUrl}/data-quality/issues/${issueId}/resolve`, request);
  }

  /**
   * Get MPI performance metrics
   */
  getMetrics(startDate?: string, endDate?: string): Observable<MPIMetrics> {
    let params = new HttpParams();
    if (startDate) params = params.set('startDate', startDate);
    if (endDate) params = params.set('endDate', endDate);

    return this.http.get<MPIMetrics>(`${this.baseUrl}/metrics`, { params });
  }

  /**
   * Get accuracy trends over time
   */
  getAccuracyTrends(days: number = 30): Observable<MPITrendData> {
    const params = new HttpParams().set('days', days.toString());
    return this.http.get<MPITrendData>(`${this.baseUrl}/trends`, { params });
  }
}
